pageserver: fix sharding emitting empty image layers during compaction

jcsp opened the pull request:
## Problem

Sharded tenants would sometimes try to write empty image layers during compaction: this was more noticeable on larger databases.
- https://github.com/neondatabase/neon/issues/6755

**Note to reviewers: the last commit is a refactor that de-intents a whole block, I recommend reviewing the earlier commits one by one to see the real changes**

## Summary of changes

- Fix a case where when we drop a key during compaction, we might fail to write out keys (this was broken when vectored get was added)
- If an image layer is empty, then do not try and write it out, but leave `start` where it is so that if the subsequent key range meets criteria for writing an image layer, we will extend its key range to cover the empty area.
- Add a compaction test that configures small layers and compaction thresholds, and asserts that we really successfully did image layer generation.  This fails before the fix.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

arpad-m commented:
> Note to reviewers: the last commit is a refactor that de-intents a whole block, I recommend reviewing the earlier commits one by one to see the real changes

btw there is an "ignore whitespace" option that you can toggle in the "changed files" view on github. Or you can just add a `w=1` query parameter to the URL: https://github.com/neondatabase/neon/pull/6776/files?diff=split&w=1

arpad-m commented on compute_tools/src/config.rs:
```diff
@@ -51,6 +51,9 @@ pub fn write_postgres_conf(
         writeln!(file, "neon.pageserver_connstring={}", escape_conf_value(s))?;
     }
+    if let Some(stripe_size) = spec.shard_stripe_size {
+        writeln!(file, "neon.stripe_size={}", stripe_size)?;
```
```suggestion
        writeln!(file, "neon.stripe_size={stripe_size}")?;
```

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -3274,90 +3274,111 @@ impl Timeline {
                 let image_layer = image_layer_writer.finish(self).await?;
                 image_layers.push(image_layer);
+            } else {
+                // Special case: the image layer may be empty if this is a sharded tenant and the
+                // partition does not cover any keys owned by this shard.  In this case, to ensure
+                // we don't leave gaps between image layers, leave `start` where it is, so that the next
+                // layer we write will cover the key range that we just scanned.
```
Wondering if this could interefere with some heuristics elsewhere (placeholder).

jcsp replied:
Right, the logic here is a bit twisted because of the function having the optional force flag:
- If `force` is set then we'll never leave gaps
- If it isn't, then an empty layer's key range will be merged into the subsequent layer _if_ the subsequent layer is selected for image layer generation.  Else when we hit the next layer, we'll advance `start` anyway.

koivunej commented on pageserver/src/tenant/timeline.rs:
```diff
@@ -3274,90 +3274,111 @@ impl Timeline {
+                // we don't leave gaps between image layers, leave `start` where it is, so that the next
+                // layer we write will cover the key range that we just scanned.
+                tracing::debug!(
+                    "create_image_layers: no data in range {}-{}",
```
```suggestion
                    "no data in range {}-{}",
```

It is already within a span called `create_image_layers`.

VladLazar commented on test_runner/regress/test_sharding.py:
```diff
@@ -255,3 +258,75 @@ def shards_on_disk(shard_ids):
+        image_layer_sizes = {}
+        for layer in layer_map.historic_layers:
+            if layer.kind == "Image":
+                image_layer_sizes[layer.layer_file_name] = layer.layer_file_size
```
Could also assert on the layer file size being non zero.

jcsp replied:
Sensible.  added in 6b704d706828686e8c5cedf38d807ca941ccdc2e

jcsp commented:
I've reverted the test in this PR and moved it to a followup PR: https://github.com/neondatabase/neon/pull/6788

To make that test run reliably, I need to fix an unrelated issue with disk_consistent_lsn, and the fix is quite delicate so don't want to rush it here.
